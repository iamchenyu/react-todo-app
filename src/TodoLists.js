import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

const INITIAL_STATE = [
  { id: uuid(), text: "Feed Dogs", finished: true },
  { id: uuid(), text: "Go To Gym", finished: false },
];

const TodoLists = () => {
  //   if (!localStorage.getItem("todos")) {
  //     console.log("here1");
  //     localStorage.setItem("todos", JSON.stringify(INITIAL_STATE));
  //   } else {
  //     console.log("here2");
  //     console.log(todos);
  //     localStorage.setItem("todos", JSON.stringify(todos));
  //   }

  const [todos, setTodos] = useState(INITIAL_STATE);

  const deleteHandler = (removeId) => {
    setTodos(todos.filter((todo) => todo.id !== removeId));
  };

  const addHandler = (todo) => {
    setTodos((todos) => [
      ...todos,
      { id: uuid(), text: todo, finished: false },
    ]);
  };

  const editHandler = (editId, editText) => {
    // Old version
    // setTodos((todos) => {
    //   let todosCopy = [...todos];
    //   let todo = todosCopy.find((todo) => todo.id === editId);
    //   todo.text = editText;
    //   return todosCopy;
    // });

    // New version
    setTodos(
      todos.map((todo) =>
        todo.id === editId ? { ...todo, text: editText } : todo
      )
    );
  };

  const renderTodos = todos.map(({ id, text, finished }) => (
    <Todo
      key={id}
      id={id}
      text={text}
      finished={finished}
      onDelete={deleteHandler}
      onEdit={editHandler}
    />
  ));

  return (
    <div className="TodoLists">
      <NewTodoForm onAdd={addHandler} />
      {renderTodos}
    </div>
  );
};

export default TodoLists;
