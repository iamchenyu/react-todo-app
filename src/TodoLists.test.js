import { render, fireEvent, getByText } from "@testing-library/react";
import TodoLists from "./TodoLists";

it("renders without crashing", () => {
  render(<TodoLists />);
});

it("matches the snapshot", () => {
  const { asFragment } = render(<TodoLists />);
  expect(asFragment()).toMatchSnapshot();
});

it("removes the todo", () => {
  const { queryByText } = render(<TodoLists />);
  const todo = queryByText("Feed Dogs");
  expect(todo).toBeInTheDocument();

  const deleteButton = todo.nextElementSibling;
  fireEvent.click(deleteButton);
  expect(todo).not.toBeInTheDocument();
});

it("adds a new todo", () => {
  const { queryByText, getByTestId } = render(<TodoLists />);
  expect(queryByText("Do Homework")).not.toBeInTheDocument();

  const input = getByTestId("new-todo-form-input");
  const button = getByTestId("new-todo-form-button");

  fireEvent.change(input, { target: { value: "Do Homework" } });

  fireEvent.click(button);

  expect(queryByText("Do Homework")).toBeInTheDocument();
  expect(input).toHaveValue("");
});

it("edits todo", () => {
  const { queryByText, queryByTestId } = render(<TodoLists />);
  const todo = queryByText("Feed Dogs");
  const editButton = todo.nextElementSibling.nextElementSibling;

  let editForm = queryByTestId("Feed Dogs");
  expect(editForm).toBeNull();

  fireEvent.click(editButton);

  editForm = queryByTestId("Feed Dogs");
  expect(editForm).toBeInTheDocument();

  const input = editForm.firstChild;
  const button = editForm.lastChild;

  fireEvent.change(input, {
    target: { value: "Do Homework" },
  });

  fireEvent.click(button);

  expect(queryByText("Feed Dogs")).not.toBeInTheDocument();
  expect(queryByText("Do Homework")).toBeInTheDocument();
});
