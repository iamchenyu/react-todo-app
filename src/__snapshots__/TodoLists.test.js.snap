// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoLists"
  >
    <form
      class="NewTodoForm"
      data-testid="new-todo-form"
    >
      <input
        data-testid="new-todo-form-input"
        name="todo"
        placeholder="what i need to do next..."
        type="text"
        value=""
      />
      <button
        data-testid="new-todo-form-button"
      >
        Submit
      </button>
    </form>
    <div
      class="Todo"
    >
      <input
        checked=""
        type="checkbox"
      />
      <span
        style="text-decoration: line-through;"
      >
        Feed Dogs
      </span>
      <button>
        X
      </button>
      <button>
        Edit Todo
      </button>
    </div>
    <div
      class="Todo"
    >
      <input
        type="checkbox"
      />
      <span
        style="text-decoration: none;"
      >
        Go To Gym
      </span>
      <button>
        X
      </button>
      <button>
        Edit Todo
      </button>
    </div>
  </div>
</DocumentFragment>
`;
